### PR TITLE
Fix COMP rewards bugs

### PR DESCRIPTION
**Pull request by TRiLeZ**
This PR fixes two separate but related bugs involving COMP accrual.

- Bug 1: borrow/supply state index not properly set when an active market gets its compSpeed set for the first time causing the market to get bricked.
- Bug 2: extra interest is accrued in markets where their compSpeed is non-zero, set to zero, then set to non-zero again with time and market activity in-between.

Note: Test cases need to be written to ensure both of these bugs never occur again (should be put in spec/scenario/Flywheel/Flywheel.scen).

### CODE REVIEW

File: contracts/Comptroller.sol [hunk 1]
```diff
@@ -1076,19 +1076,21 @@ contract Comptroller is ComptrollerV5Storage, ComptrollerInterface, ComptrollerE
...
 
-            if (compSupplyState[address(cToken)].index == 0 && compSupplyState[address(cToken)].block == 0) {
+            if (compSupplyState[address(cToken)].index == 0) {
                 compSupplyState[address(cToken)] = CompMarketState({
                     index: compInitialIndex,
                     block: safe32(getBlockNumber(), "block number exceeds 32 bits")
                 });
-            }
+            } else // Update block number to ensure extra interest is not accrued during the prior period
+                compSupplyState[address(cToken)].block = safe32(getBlockNumber(), "block number exceeds 32 bits");
```
Comment (jflatow) [inline contracts/Comptroller.sol:1085]: These `else` clauses should have braces `{ ... }`

**Review by jflatow — COMMENTED** [review]
Good job finding a bug in the Gauntlet changes. Sadly, these changes make a complicated function even more complicated. I think there is a bunch more work to be done to show that these are the correct set of fixes (testing wise), and we should consider if there is something that can be done to simplify this code while also fixing it (otherwise we've introduced a bunch of new untested branches and my own faith in this code continues to drop). Also I think whoever does this work properly should receive a bunch of COMP.

**Comment by jflatow** [comment]
Also, this reminds me of the initialization issue https://github.com/compound-finance/compound-protocol/issues/40, which I remember was particularly tricky to have avoided but don't recall all the reasons why. I think whoever fixes this should have an understanding of what went wrong then too, and how this is different, and what changed in this code in between.

**Comment by TRiLeZ** [comment]
> Good job finding a bug in the Gauntlet changes. Sadly, these changes make a complicated function even more complicated. I think there is a bunch more work to be done to show that these are the correct set of fixes (testing wise), and we should consider if there is something that can be done to simplify this code while also fixing it (otherwise we've introduced a bunch of new untested branches and my own faith in this code continues to drop). Also I think whoever does this work properly should receive a bunch of COMP.

Thanks for the reply. I agree that additional test cases should be written to ensure that the problem is entirely fixed and so it doesn't happen again. elee said he would help in this regard.

I disagree with simplifying the code at the same time. The more changes, the more likely that new bugs will be introduced. I think the bug should be fixed first with a minimal amount of changes, then any simplifications can be done at a later time with even more review and testing.

**Comment by TRiLeZ** [comment]
> Also, this reminds me of the initialization issue #40, which I remember was particularly tricky to have avoided but don't recall all the reasons why. I think whoever fixes this should have an understanding of what went wrong then too, and how this is different, and what changed in this code in between.

This patch fixes the initialization issue. Getty, elee, and I worked together on this and we all have an understanding of what went wrong, why it happened, and how these changes fix the issue. We've written a report here: [Comptroller compSpeed bug](https://www.comp.xyz/t/comptroller-compspeed-bug/2111).

**Comment by jflatow** [comment]
> > Also, this reminds me of the initialization issue #40, which I remember was particularly tricky to have avoided but don't recall all the reasons why. I think whoever fixes this should have an understanding of what went wrong then too, and how this is different, and what changed in this code in between.
> 
> This patch fixes the initialization issue. Getty, elee, and I worked together on this and we all have an understanding of what went wrong, why it happened, and how these changes fix the issue. We've written a report here: [Comptroller compSpeed bug](https://www.comp.xyz/t/comptroller-compspeed-bug/2111).

There were changes to this code since issue #40 was initially filed, I haven't seen any discussion about that?

**Comment by elee1766** [comment]
> 
> 
> Good job finding a bug in the Gauntlet changes. Sadly, these changes make a complicated function even more complicated. I think there is a bunch more work to be done to show that these are the correct set of fixes (testing wise), and we should consider if there is something that can be done to simplify this code while also fixing it (otherwise we've introduced a bunch of new untested branches and my own faith in this code continues to drop). Also I think whoever does this work properly should receive a bunch of COMP.


i mentioned to trilez in our gc chat that it all looked a bit complicated, but we felt it was best to not touch the contracts too much for the purposes of patching the error and moving forward as worse is better for now.

i'm more than happy to go through and rewrite all this logic in a cleaner fashion, I personally feel it would be good in the long run to fix up the code instead of trying to throw water off board with buckets.

**Comment by jflatow** [comment]
> > Good job finding a bug in the Gauntlet changes. Sadly, these changes make a complicated function even more complicated. I think there is a bunch more work to be done to show that these are the correct set of fixes (testing wise), and we should consider if there is something that can be done to simplify this code while also fixing it (otherwise we've introduced a bunch of new untested branches and my own faith in this code continues to drop). Also I think whoever does this work properly should receive a bunch of COMP.
> 
> i mentioned to trilez in our gc that the contract looked a bit messier than it should, but we felt it was best to not touch the contracts too much for the purposes of patching the error and moving forward as worse is better for now.
> 
> i'm more than happy to go through and rewrite all this logic in a cleaner fashion, I personally feel it would be good in the long run to fix up the code instead of trying to throw water off board with buckets.

Yeah I kinda agree, I'm just concerned there's actually some more subtle issues here that aren't being addressed. There are a lot of branches now so its really hard to follow and I don't believe we have coverage of all the cases that can occur anymore
